src/Reflection/InitializerExprTypeResolver.php

```diff
@@ -497,6 +514,90 @@ public function getArrayType(Expr\Array_ $expr, callable $getTypeCallback): Type
 		return $arrayType;
 	}
 
+	/**
+	 * @param callable(Expr): Type $getTypeCallback
+	 *
+	 * @throws OversizedConstantArrayTypeException
+	 */
+	private function builtDegradedConstantArray(Array_ $node, callable $getTypeCallback): Type
+	{
+		$isList = true;
+		$isEmptyArray = true;
+		$isKeysNonEmptyString = true;
+		$isValuesNonEmptyString = true;
+
+		$valueTypes = [];
+		$itemKeyTypes = [];
+		foreach ($node->items as $item) {
+			if ($item === null) {
+				continue;
+			}
+
+			$isEmptyArray = false;
+
+			$key = $item->key;
+			if ($key !== null) {
+				$isList = false;
+
+				$itemKeyType = $getTypeCallback($key);
+				if (!$itemKeyType instanceof ConstantScalarType || $key instanceof ClassConstFetch) {
+					throw new OversizedConstantArrayTypeException();
+				}
+				if (!$itemKeyType->isNonEmptyString()->yes()) {
+					$isKeysNonEmptyString = false;
+				}
+				$generalizedType = $itemKeyType->generalize(GeneralizePrecision::lessSpecific());
+				$itemKeyTypes[get_class($generalizedType)] = $generalizedType; // de-duplicate values
+			} else {
+				$isKeysNonEmptyString = false;
+			}
+
+			$value = $item->value;
+			if ($value instanceof Array_) {
+				$valueTypes[] = $this->builtDegradedConstantArray($value, $getTypeCallback);
+				continue;
+			}
+
+			$itemValueType = $getTypeCallback($value);
+			if (!$itemValueType instanceof ConstantScalarType || $value instanceof ClassConstFetch) {
```

Comment:
I think we could also add support for arrays where const arrays call methods/functions as long as the return type is defined.

I did not yet work on it, as I think we are still discussing whether the change itself is acceptable.